plugins/amqp_trace_plugin/include/eosio/amqp_trace_plugin/amqp_handler.hpp

```diff
@@ -0,0 +1,178 @@
+#pragma once
+
+#include <eosio/chain/thread_utils.hpp>
+#include <fc/log/logger.hpp>
+
+#include <amqpcpp.h>
+#include <amqpcpp/libboostasio.h>
+#include <amqpcpp/linux_tcp.h>
+
+
+namespace eosio {
+
+/// Designed to work with internal io_service running on a dedicated thread.
+/// All public interface methods can be called from any thread.
+class amqp {
+public:
+   // called from amqp thread on errors
+   using on_error_t = std::function<void(const std::string& err)>;
+   // called from amqp thread on consume of message
+   // return empty optional for no ack/reject of message
+   //        true for ack, false for reject
+   using on_consume_t = std::function<std::optional<bool>(const char* buf, size_t s)>;
+
+   /// @param address AMQP address
+   /// @param name AMQP routing key
+   /// @param on_err callback for errors, called from amqp thread, can be nullptr
+   /// @param on_consume callback for consume on routing key name, called from amqp thread, null if no consume needed
+   amqp( const std::string& address, std::string name, on_error_t on_err, on_consume_t on_consume = nullptr )
+         : thread_pool_( "ampq", 1) // amqp is not thread safe, use only one thread
+         , name_( std::move( name ) )
+         , on_error_( std::move( on_err ) )
+         , on_consume_( std::move( on_consume ) )
+   {
+      AMQP::Address amqp_address( address );
+      ilog( "Connecting to AMQP address ${a} - Queue: ${q}...", ("a", std::string( amqp_address ))( "q", name_ ) );
+
+      handler_ = std::make_unique<amqp_handler>( *this, thread_pool_.get_executor() );
+      connection_ = std::make_unique<AMQP::TcpConnection>( handler_.get(), amqp_address );
```

Comment:
The handler is safe to construct in the main thread (at least as long as it happens before the construction of the connection).  The connection is definitely not safe to construct here.  `AMQP::TcpConnection` begins issuing async operations before it is fully constructed.

Reply:
Moved to amqp thread.